core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4509,6 +4510,7 @@ public static void _doScript(StaplerRequest req, StaplerResponse rsp, RequestDis
             }
 
             try {
+                LOGGER.log(Level.INFO, req.getRemoteAddr() + "|" + Jenkins.getAuthentication().getName() + "|ScriptConsole|" + Base64.getEncoder().encodeToString(text.getBytes()));
```

Comment:
Log messages in Jenkins are typically intended to be human-readable. I vaguely remember pushback on the multiline messages with the password when setting up a new instance, but there's no reason even with base64 script this shouldn't be a human readable message.

How about this?

```suggestion
                LOGGER.log(Level.INFO, "Administrative Groovy script by " + Jenkins.getAuthentication().getName() + " from " + req.getRemoteAddr() + ": (base64) " + Base64.getEncoder().encodeToString(text.getBytes()));
```

Why is this not logging in `RemotingDiagnostics` where the actual call happens?

Reply:
Yes, the use case is to detect breaches when remote logging is set up. However, the logging level needs to be at least INFO so that it's sent to the logs by default.

Imagine a company that has a golden image with the remote logging set up at the OS level, then a developer downloads and installs the latest version of Jenkins. I want them to be protected by default. Otherwise, that would involve modifying the Jenkins install to set up the log listeners for the lower-severity logs. If security requires additional configuration by the user, that makes it more likely to get messed up.

As for the log format, it doesn't matter too much to me as long as the username, IP, event name, and script is in there in a one-line format that also takes into account the potential for log injection (hence the base64).

I am totally new to reviewing Jenkins code, so I found a place that was able to get all of the info I needed (IP, username, script contents). Ideally this is a chokepoint that will always be triggered, so that hitting it from API vs GUI doesn't bypass the logging.

I have some other questions about improving Jenkins' security logging. What's the best way to reach you?

Reply:
> Imagine a company that has a golden image with the remote logging set up at the OS level

I haven't set this up before, but I'm surprised it wouldn't be possible to specify a more verbose log level for where you define the remote logging setup. Are you sure that's the case?

> I have some other questions about improving Jenkins' security logging. What's the best way to reach you?

If it's about implementation feedback, PRs. Improvements/Features/Non sensitive issues related: File a JENKINS issue and ping me there. Sensitive (e.g. possible security vulnerability): File a SECURITY issue.

Reply:
> I haven't set this up before, but I'm surprised it wouldn't be possible to specify a more verbose log level for where you define the remote logging setup. Are you sure that's the case?

It's my understanding that Jenkins logs only show up on STDOUT if they're INFO or higher. Otherwise you need to go into Jenkins to set up the listener to capture stuff. I'm double checking with my logging guy, but if the log entry never gets written to disk, then we can't forward the log. So for example, if you throw jenkins.war at Tomcat, the Jenkins STDOUT would be in catalina.out, and we can "live stream" the lines being written to that file as they are written, and store them remotely.

But if Jenkins doesn't output it by default, defenders can't see it. This is why I'm pushing for INFO level.

Reply:
Write a `java.util.logging` configuration properties file and add a reference to it to the Jenkins invocation. JENKINS-57888, while not about that, gives some examples of these in its description to get you started.

Reply:
@daniel-beck 

Unfortunately, the issue with that approach is that it modifies the Jenkins install/config, which is outside the use case (and associated threat model) of a prebuilt image with OS-level configuration to gain read access to arbitrary log files (that users then later install the software they need--not having a unique golden image for every product). Doing what you propose involves adding write access to arbitrary files so that the defender can mess with installs (and effectively RCE to relaunch the process under the new config). I'm not really down to have a backdoor in my golden images (not even for myself), and I don't want this hardening to rely on the user (or involve them at all), at least not for a product like Jenkins. We should strive for security by default.

I think a lot of companies are in the same boat of their defenders (or, at least, the "logging people") only having read access. A defender that has write access could, for instance, simply patch Jenkins and wouldn't need to do a public pull request. Write access also adds liability. I don't want to be on the hook for breaking their setup.

I think INFO is a fair ask considering this activity should be relatively rare (it was, after all, designed for troubleshooting) so it shouldn't be too spammy (if that's the concern). Similarly interesting activity, like logging into the CLI via SSH, is already INFO.

Hope you understand where I'm coming from. If you still have concerns on your side, please let me know. I wanna work this out.

Reply:
IIUC in

> golden image with the remote logging set up at the OS level

you're already changing your image to grab stdout and send it elsewhere; it's not like you can do that with our unmodified stock images. How is a change to `/etc/defaults/jenkins` or equivalent during image building different?

Reply:
In this case, the golden image itself is not grabbing STDOUT. When you throw jenkins.war at Tomcat, Tomcat sends STDOUT to catalina.out. The level of access that many defenders have is to specify which files on disk to live-stream a copy of and send that data remotely. This stuff can be changed after the fact ("oh they installed Tomcat? Let's start grabbing Tomcat logs"), because the defenders have read access to arbitrary files.

The golden image is per operating system, not by application.

Reply:
Thanks for the explanation.

TBH I'd like to see more demand for something like this before it gets implemented.

One org with restrictive policies doesn't seem sufficient motivation to do a change like this, especially since I know that many setups heavily script Jenkins through these APIs, and getting the same giant periodic scripts logged every time interval would be unnecessary log spam.

Reply:
(FTR, poking a few folks out of band for their thoughts.)

Reply:
Sure, I'm part of one org. But I'm certainly not the only one with this demand.
https://issues.jenkins-ci.org/browse/JENKINS-41516

> Unless there's an external append-only log of sorts configured, no improvements here are likely to substantially improve the situation. And once you need an admin to take action to protect themselves, it's easier to secure Jenkins properly.

We have external append-only logs. Please help me and other others substantially improve this situation. Modifying the Jenkins installation is requiring an admin to take actions to protect themselves. Securing Jenkins properly includes having security-relevant logs available by default.

Reply:
> Modifying the Jenkins installation is requiring an admin to take actions to protect themselves.

My suggestion is to have preconfigured custom images that send the logs to external append only loggers. That's certainly possible today in general using the tools I mentioned. That your org's chosen approach to external logging doesn't work is unfortunate but doesn't invalidate the approach in general.

Reply:
Thanks for poking additional people to join the discussion. I think this has the potential to help a lot of people, and I'm sure that together we can figure something out that makes sense for all parties.

Reply:
@daniel-beck any additional thoughts from your colleagues?

Reply:
Sorry for only responding now.

I didn't get useful results from my first contacts back then and was busy in the mean time.